### PR TITLE
VS30: Support Bson Attributes

### PR DESCRIPTION
**Pull request by Ravi-Raghavan**
VS30: Support Bson Attributes

### CODE REVIEW

File: src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs [hunk 1]
```diff
@@ -16,6 +16,46 @@ namespace MongoDB.Analyzer.Core;
...
+               !attribute.NamedArguments.Where(a => !IsSupportedAttributeArgumentKind(a.Value)).EmptyOrNull())
+            {
+                continue;
+            }
+
+            foreach (var constructorArgument in attribute.ConstructorArguments)
+            {
+                var literalToken = SyntaxFactoryUtilities.GetConstantValueToken(constructorArgument.Value);
+                var literalSyntaxKind = literalToken.Kind() == SyntaxKind.StringLiteralToken ? SyntaxKind.StringLiteralExpression : SyntaxKind.NumericLiteralExpression;
+                argumentList = argumentList.Add(SyntaxFactory.AttributeArgument(SyntaxFactory.LiteralExpression(literalSyntaxKind, literalToken)));
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs:38]: `argumentList.AddRange(attribute.ConstructorArguments.Select(ToAttributeArgument));`

File: src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs [hunk 2]
```diff
@@ -16,6 +16,46 @@ namespace MongoDB.Analyzer.Core;
...
+                var argumentNameIdentifier = SyntaxFactory.IdentifierName(namedArgument.Key);
+                var literalToken = SyntaxFactoryUtilities.GetConstantValueToken(namedArgument.Value.Value);
+                var literalSyntaxKind = literalToken.Kind() == SyntaxKind.StringLiteralToken ? SyntaxKind.StringLiteralExpression : SyntaxKind.NumericLiteralExpression;
+                argumentList = argumentList.Add(SyntaxFactory.AttributeArgument(SyntaxFactory.NameEquals(argumentNameIdentifier), SyntaxFactory.NameColon(argumentNameIdentifier), SyntaxFactory.LiteralExpression(literalSyntaxKind, literalToken)));
+            }
+
+            attributeList = attributeList.Add(SyntaxFactory.Attribute(SyntaxFactory.IdentifierName(attribute.AttributeClass.Name), SyntaxFactory.AttributeArgumentList(argumentList)));
+        }
+
+        memberDeclaration = !attributeList.EmptyOrNull() ? memberDeclaration.AddAttributeLists(SyntaxFactory.AttributeList(attributeList)) : memberDeclaration;
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs:52]: Any()?
minor: var result = attributeList.Any() ? ... : ...

File: tests/MongoDB.Analyzer.Tests/Linq/LinqTests.cs [hunk 3]
```diff
@@ -65,4 +65,8 @@ public sealed class LinqTests : DiagnosticsTestCasesRunner
     [DataTestMethod]
     [CodeBasedTestCasesSource(typeof(LinqQuerySyntax))]
     public Task QuerySyntax(DiagnosticTestCase testCase) => VerifyTestCase(testCase);
+
+    [DataTestMethod]
+    [CodeBasedTestCasesSource(typeof(LinqBsonAttributes))]
+    public Task BsonAttributes(DiagnosticTestCase testCase) => VerifyTestCase(testCase);
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests/Linq/LinqTests.cs:71]: Alphabetical order.

File: src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs [hunk 4]
```diff
@@ -16,6 +16,46 @@ namespace MongoDB.Analyzer.Core;
 
 internal static class SyntaxNodeExtensions
 {
+    public static MemberDeclarationSyntax AddBsonAttributes(this MemberDeclarationSyntax memberDeclaration, IEnumerable<AttributeData> attributes)
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs:19]: Check if `T AddBsonAttributes<T>(this T memberDeclaration, ...) where T : MemberDeclarationSyntax` works.
might eliminate the need for casting as FieldDeclarationSyntax, and ensure type safety.

File: src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs (hunk 2 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs:52]: Done

File: src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs (hunk 4 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs:19]: Done

File: tests/MongoDB.Analyzer.Tests.Common/DataModel/University.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+using System;
+using System.Collections.Generic;
+using MongoDB.Bson.Serialization.Attributes;
+
+namespace MongoDB.Analyzer.Tests.Common.DataModel
+{
+    public class University
+    {
+        [BsonDefaultValue("Princeton University")]
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common/DataModel/University.cs:23]: only reason i included it was to ensure that having it didn't destroy existing functionality

File: tests/MongoDB.Analyzer.Tests.Common/DataModel/University.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            Name = name;
+        }
+
+        [BsonRequired]
+        public string Name { get; set; }
+
+        [BsonElement("StudentAge")]
+        public int Age { get; set; }
+
+
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common/DataModel/University.cs:62]: Done

File: src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs [hunk 7]
```diff
@@ -16,6 +16,46 @@ namespace MongoDB.Analyzer.Core;
...
+    {
+        attributes = attributes?.Where(a => a.AttributeClass.IsBsonAttribute());
+        var attributeList = new SeparatedSyntaxList<AttributeSyntax>();
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs:22]: Done

File: src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs [hunk 8]
```diff
@@ -16,6 +16,46 @@ namespace MongoDB.Analyzer.Core;
...
+
+        foreach (var attribute in attributes)
+        {
+            var argumentList = new SeparatedSyntaxList<AttributeArgumentSyntax>();
+
+            if (!attribute.ConstructorArguments.Where(a => !IsSupportedAttributeArgumentKind(a)).EmptyOrNull() ||
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs:28]: Done

File: src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs (hunk 1 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/SyntaxNodeExtensions.cs:38]: Done

**Pull request closed without merging**